[Self-service error-codes] Add introductory section to self-service error codes page. (#11712)
digital-asset/daml: docs/source/error-codes/self-service/index.rst

.. toctree::
   :hidden:

Overview
*********

TODO: Failure vs. error.

.. _gRPC: https://grpc.github.io/grpc/core/md_doc_statuscodes.html

Participant API server has been using a combination of gRPC_
status codes and custom messages to signal erroneous or unexpected conditions.
This approach remains unchanged in principle while self-service error codes aim at
enhancing it by providing:

- improved consistency of the returned errors across API endpoints,

- richer message format with clearly distinguished machine readable parts to facilitate
  automated error handling strategies,

- complete inventory of all error codes with an explanation, suggested resolution and
  more (as documented in this section).

The goal is to enable users, developers and operators to act upon encountered
errors in a self-service manner such that the need to escalate the problem

> or file a ticket occurs occurs infrequently.

Reviewer: ```suggestion
or file a ticket occurs infrequently.
```

Feature flag
************
You can enable self-service error-codes by specifying ``--use-self-service-error-codes``
from command line.

By default self-service error codes are turned off.

Anatomy of an error code
*****************************

TODO Error categories:

 - what they are,

 - how to make use of them,

 - provide a list of all error categories.

Error message
===================
.. code-block:: java

>     <ERROR_CODE_ID>(<CATEGORY_ID>,<CORRELATION_ID>:<HUMAN_READABLE_MESSAGE>

Reviewer: The parenthesis need to be closed

The constituent parts are:

  - ``<ERROR_CODE_ID>`` - a unique non empty string containing at most 63 characters:
    upper-cased letters, underscores or digits.
  - ``<CATEGORY_ID>`` - a small integer identifying the corresponding error categories.

>   - ``<CORRELATION_ID>`` - a string helpful at identifying originating request.

Reviewer: ```suggestion
  - ``<CORRELATION_ID>`` - a string helpful in identifying originating request.
```

>     ``0`` if no correlation id is given. TODO: Are there limits to its size?

Reviewer: We take this value from the submission-id (where provided). The limit is thus the same as for submission-id. We only say there that it must be a valid ledger string and provide a regex for it no mention of the size. Typically this will be a UUID

     - The purpose of the correlation-id is to allow a user to clearly identify the request,
       such that the operator can lookup any log information associated with this error.
       TODO: Make sure that's true.

>   - ``:`` - a colon character that server as a separator for the machine and human readable parts.

Reviewer: ```suggestion
  - ``:`` - a colon character that serves as a separator for the machine and human readable parts.
```

  - ``<HUMAN_READABLE_MESSAGE>`` - message targeting a human reader.

> A concrete example how a message might look like this:

Reviewer: ```suggestion
In a concrete example a message might look like this:
```

.. code-block:: java

    TRANSACTION_NOT_FOUND(11,12345): Transaction not found, or not visible.

Error metadata
====================

> In addition error response contain the following metadata:

Reviewer: ```suggestion
In addition, an error response contains the following metadata:
```

- mandatory ``com.google.rpc.ErrorInfo`` containing <ERROR_CODE>,

- mandatory ``com.google.rpc.RequestInfo`` containing trace id,

- optional ``com.google.rpc.RetryInfo`` containing retry interval,

- optional ``com.google.rpc.ResourceInfo`` containing key value pair
  of the form: <RESOURCE_NAME>: <RESOURCE_IDENTIFIER>,
  e.g. ""TRANSACTION_ID", "tId12345".

>   TODO: Document all possible resource names?

Reviewer: Just say that the resources are implementation specific and may vary from ledger to ledger

Logging
********

> TODO: Excerpt copied form Canton docs verbatim:

Reviewer: In this section I would say that most of the errors are printed at their creation time at a level appropriate for their error category. The information contained in the logfile is consistent with the data returned over the gRPC.

Then I would dissect an example.

Author: Let's talk about this offline

+------------------------------------------------+------------+--------------------+
|BackgroundProcessDegradationWarning             |13          |N/A                 |
+------------------------------------------------+------------+--------------------+

Anatomy of an Error
==========================

Errors returned to users are represented as instances of standard StatusRuntimeException_.
As such they contain a `gRPC status code`_, a description and additional machine readable information
represented in the `rich gRPC error model`_.

Error Description
--------------------------

We use the `standard gRPC description`_ that additionally adheres to our custom message format:

.. code-block:: java

    <ERROR_CODE_ID>(<CATEGORY_ID>,<TRUNCATED_CORRELATION_ID>):<HUMAN_READABLE_MESSAGE>

The constituent parts are:

  - ``<ERROR_CODE_ID>`` - a unique non empty string containing at most 63 characters:
    upper-cased letters, underscores or digits.
    Identifies corresponding error code id.

  - ``<CATEGORY_ID>`` - a small integer identifying the corresponding error category.

  - ``<TRUNCATED_CORRELATION_ID>`` - a string aimed at identifying originating request.

>     Contains at most 8 characters of the original correlation id.

Reviewer 2: This is a prefix of the submission id

Author: Will add in a follow-up PR as now the CI is green. Wdyt?

.. _rich gRPC error model: https://cloud.google.com/apis/design/errors#error_details
.. _standard gRPC description: https://grpc.github.io/grpc-java/javadoc/io/grpc/Status.html#getDescription--

The majority of the errors are a result of some request processing.
They are logged and returned to the user as a failed gRPC request
using the standard StatusRuntimeException_.
As such this approach remains unchanged in principle while self-service error codes aim at
enhancing it by providing:

- improved consistency of the returned errors across API endpoints,

- richer error payload format with clearly distinguished machine readable parts to facilitate
  automated error handling strategies,

- complete inventory of all error codes with an explanation, suggested resolution and
  other useful information.

The goal is to enable users, developers and operators to act on the encountered
errors in a self-service manner, either in an automated-way or manually.

Feature Flag
---------------------------

You can enable self-service error-codes by specifying ``--use-self-service-error-codes``
from command line.

> By default self-service error codes are turned off.

Reviewer 2: It will be enabled by default. Let's not change this in this PR, though. I'll do it in a subsequent PR

Glossary
---------------------------

Error
        Represents an occurrence of a failure.
        Consists of:

        - an `error code id`,

        - a `gRPC status code`_ (determined by its error category),

        - an `error category`,

>         - a `correlation id`.

Reviewer 2: Please do refer in these docs that the correlation id is generally populated by the submission id.

Author: Ok

Author: Follow-up PR: https://github.com/digital-asset/daml/pull/11728
